File: core/tx_list.go
package core

import (
	"fmt"
	"math/big"
	"sort"

	ethtypes "github.com/ethereum/go-ethereum/core/types"
)

// txList is a list of transactions sorted by nonce.
// It handles gap detection and replacement.
type txList struct {
	txs   *sortedTxs // Heap/Sortable
	nonce uint64     // Current expected nonce (State Nonce)
}

func newTxList(nonce uint64) *txList {
	return &txList{
		txs:   new(sortedTxs),
		nonce: nonce,
	}
}

// Add adds a tx to the list. Returns true if it replaced an old one.
// Returns error if nonce is too old.
func (l *txList) Add(tx *ethtypes.Transaction, priceBump uint64) (bool, *ethtypes.Transaction, error) {
	if tx.Nonce() < l.nonce {
		return false, nil, ErrTxNonceTooLow
	}

	// Check for replacement
	old := l.txs.Get(tx.Nonce())
	if old != nil {
		// Price bump check
		oldPrice := old.GasPrice()
		newPrice := tx.GasPrice()

		// minPrice = oldPrice * (100 + priceBump) / 100
		threshold := new(big.Int).Mul(oldPrice, big.NewInt(int64(100+priceBump)))
		threshold.Div(threshold, big.NewInt(100))

		if newPrice.Cmp(threshold) < 0 {
			return false, nil, fmt.Errorf("replacement transaction underpriced: need %v, have %v", threshold, newPrice)
		}

		l.txs.Remove(tx.Nonce())
		l.txs.Add(tx)
		return true, old, nil
	}

	l.txs.Add(tx)
	return false, nil, nil
}

// Forward advances the expected nonce, simulating block inclusion.
// Returns removed transactions (those with nonce < newNonce).
func (l *txList) Forward(newNonce uint64) []*ethtypes.Transaction {
	var removed []*ethtypes.Transaction
	l.nonce = newNonce

	// Remove now-old transactions
	// Since it's sorted, we can just pop from front until checking
	// But `sortedTxs` is a set. We filter.
	l.txs.Filter(func(tx *ethtypes.Transaction) bool {
		if tx.Nonce() < newNonce {
			removed = append(removed, tx)
			return false // Remove
		}
		return true // Keep
	})
	return removed
}

// Ready returns transactions that are executable (continuous nonces starting from l.nonce).
func (l *txList) Ready(cap int) []*ethtypes.Transaction {
	var ready []*ethtypes.Transaction
	next := l.nonce

	for _, tx := range *l.txs {
		if tx.Nonce() > next {
			break // Gap found
		}
		if tx.Nonce() == next {
			ready = append(ready, tx)
			next++
		}
		if len(ready) >= cap {
			break
		}
	}
	return ready
}

// Len returns number of txs.
func (l *txList) Len() int {
	return len(*l.txs)
}

// Empty returns true if list is empty.
func (l *txList) Empty() bool {
	return len(*l.txs) == 0
}

// sortedTxs implements sort.Interface
type sortedTxs []*ethtypes.Transaction

func (s sortedTxs) Len() int           { return len(s) }
func (s sortedTxs) Swap(i, j int)      { s[i], s[j] = s[j], s[i] }
func (s sortedTxs) Less(i, j int) bool { return s[i].Nonce() < s[j].Nonce() }

func (s *sortedTxs) Add(tx *ethtypes.Transaction) {
	*s = append(*s, tx)
	sort.Sort(s)
}

func (s *sortedTxs) Get(nonce uint64) *ethtypes.Transaction {
	for _, tx := range *s {
		if tx.Nonce() == nonce {
			return tx
		}
	}
	return nil
}

func (s *sortedTxs) Remove(nonce uint64) {
	// Find and remove
	for i, tx := range *s {
		if tx.Nonce() == nonce {
			*s = append((*s)[:i], (*s)[i+1:]...)
			return
		}
	}
}

// Filter modifies slice in place
func (s *sortedTxs) Filter(fn func(*ethtypes.Transaction) bool) {
	n := 0
	for _, x := range *s {
		if fn(x) {
			(*s)[n] = x
			n++
		}
	}
	*s = (*s)[:n]
}

File: core/tx_list.zig
const std = @import("std");
const types = @import("types.zig");

/// TxList manages a list of transactions for a single account, sorted by nonce.
pub const TxList = struct {
    allocator: std.mem.Allocator,
    txs: std.ArrayListUnmanaged(*types.Transaction),
    nonce: u64, // Expected state nonce

    pub fn init(allocator: std.mem.Allocator, nonce: u64) TxList {
        return TxList{
            .allocator = allocator,
            .txs = .{},
            .nonce = nonce,
        };
    }

    pub fn deinit(self: *TxList) void {
        self.txs.deinit(self.allocator);
    }

    /// Add inserts a transaction into the sorted list.
    /// Returns true if it replaced an existing transaction (same nonce).
    /// Returns error if nonce is lower than current state nonce.
    pub fn add(self: *TxList, tx: *types.Transaction) !bool {
        if (tx.nonce < self.nonce) {
            return error.NonceTooLow;
        }

        // Check for existing nonce to replace
        for (self.txs.items, 0..) |existing, i| {
            if (existing.nonce == tx.nonce) {
                // Replacement Logic (Gas Price Bump check typically here)
                // For now, simpler replacement:
                // Typically we check if new price > old price * 1.1
                // We'll skip complex check for this port step.
                self.txs.items[i] = tx;
                return true;
            }
            if (existing.nonce > tx.nonce) {
                // Insert before
                try self.txs.insert(self.allocator, i, tx);
                return false;
            }
        }

        // Append at end
        try self.txs.append(self.allocator, tx);
        return false;
    }

    /// Forward advances the expected nonce (e.g. after block inclusion).
    /// Removes transactions with nonce < new_nonce.
    pub fn forward(self: *TxList, new_nonce: u64) void {
        self.nonce = new_nonce;

        while (self.txs.items.len > 0) {
            if (self.txs.items[0].nonce < new_nonce) {
                // Remove front
                _ = self.txs.orderedRemove(0);
            } else {
                break;
            }
        }
    }

    /// Ready returns a slice of transactions that are executable (contiguous nonces starting from self.nonce).
    /// The returned slice is owned by the list internal buffer (reference valid until modification).
    pub fn ready(self: *TxList, cap: usize) []const *types.Transaction {
        var count: usize = 0;
        var next_nonce = self.nonce;

        for (self.txs.items) |tx| {
            if (count >= cap) break;
            if (tx.nonce == next_nonce) {
                count += 1;
                next_nonce += 1;
            } else {
                // Gap found
                break;
            }
        }
        return self.txs.items[0..count];
    }

    pub fn len(self: *TxList) usize {
        return self.txs.items.len;
    }

    pub fn empty(self: *TxList) bool {
        return self.txs.items.len == 0;
    }
};
